File: swift-bridge/Sources/CoreDataBridge/PersistentStoreRequest.swift
import CoreData
import Foundation

@_cdecl("cd_persistent_store_request_get_affected_stores")
public func cdPersistentStoreRequestGetAffectedStores(_ requestPtr: UnsafeMutableRawPointer?) -> UnsafeMutableRawPointer? {
    guard let requestPtr else {
        return nil
    }
    let request: NSPersistentStoreRequest = cdBorrow(requestPtr)
    guard let stores = request.affectedStores else {
        return nil
    }
    return cdRetain(stores as NSArray)
}

@_cdecl("cd_persistent_store_request_set_affected_stores")
public func cdPersistentStoreRequestSetAffectedStores(
    _ requestPtr: UnsafeMutableRawPointer?,
    _ storePtrs: UnsafePointer<UnsafeMutableRawPointer?>?,
    _ count: Int32,
    _ outError: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?
) -> Int32 {
    do {
        guard let requestPtr else {
            throw cdBridgeNSError(code: CDR_INVALID_ARGUMENT, message: "Missing persistent store request")
        }
        let request: NSPersistentStoreRequest = cdBorrow(requestPtr)
        request.affectedStores = cdObjects(from: storePtrs, count: count)
        return CDR_OK
    } catch let error as NSError {
        cdWriteError(error, to: outError)
        return Int32(error.code)
    }
}

@_cdecl("cd_persistent_store_request_get_request_type")
public func cdPersistentStoreRequestGetRequestType(_ requestPtr: UnsafeMutableRawPointer?) -> UInt64 {
    guard let requestPtr else {
        return 0
    }
    let request: NSPersistentStoreRequest = cdBorrow(requestPtr)
    return UInt64(request.requestType.rawValue)
}

@_cdecl("cd_asynchronous_fetch_request_new")
public func cdAsynchronousFetchRequestNew(
    _ fetchRequestPtr: UnsafeMutableRawPointer?,
    _ outRequest: UnsafeMutablePointer<UnsafeMutableRawPointer?>?,
    _ outError: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?
) -> Int32 {
    do {
        guard let fetchRequestPtr else {
            throw cdBridgeNSError(code: CDR_INVALID_ARGUMENT, message: "Missing fetch request")
        }
        let fetchRequest: NSFetchRequest<NSFetchRequestResult> = cdBorrow(fetchRequestPtr)
        outRequest?.pointee = cdRetain(NSAsynchronousFetchRequest(fetchRequest: fetchRequest, completionBlock: nil))
        return CDR_OK
    } catch let error as NSError {
        cdWriteError(error, to: outError)
        return Int32(error.code)
    }
}

@_cdecl("cd_asynchronous_fetch_request_get_fetch_request")
public func cdAsynchronousFetchRequestGetFetchRequest(_ requestPtr: UnsafeMutableRawPointer?) -> UnsafeMutableRawPointer? {
    guard let requestPtr else {
        return nil
    }
    let request: NSAsynchronousFetchRequest<NSFetchRequestResult> = cdBorrow(requestPtr)
    return cdRetain(request.fetchRequest)
}

@_cdecl("cd_asynchronous_fetch_request_get_estimated_result_count")
public func cdAsynchronousFetchRequestGetEstimatedResultCount(_ requestPtr: UnsafeMutableRawPointer?) -> Int64 {
    guard let requestPtr else {
        return 0
    }
    let request: NSAsynchronousFetchRequest<NSFetchRequestResult> = cdBorrow(requestPtr)
    return Int64(request.estimatedResultCount)
}

@_cdecl("cd_asynchronous_fetch_request_set_estimated_result_count")
public func cdAsynchronousFetchRequestSetEstimatedResultCount(
    _ requestPtr: UnsafeMutableRawPointer?,
    _ estimatedResultCount: Int64
) {
    guard let requestPtr else {
        return
    }
    let request: NSAsynchronousFetchRequest<NSFetchRequestResult> = cdBorrow(requestPtr)
    request.estimatedResultCount = Int(estimatedResultCount)
}

@_cdecl("cd_managed_object_context_execute_asynchronous_fetch_request")
public func cdManagedObjectContextExecuteAsynchronousFetchRequest(
    _ contextPtr: UnsafeMutableRawPointer?,
    _ requestPtr: UnsafeMutableRawPointer?,
    _ outResult: UnsafeMutablePointer<UnsafeMutableRawPointer?>?,
    _ outError: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?
) -> Int32 {
    do {
        guard let contextPtr, let requestPtr else {
            throw cdBridgeNSError(code: CDR_INVALID_ARGUMENT, message: "Missing managed object context or asynchronous fetch request")
        }
        let context: NSManagedObjectContext = cdBorrow(contextPtr)
        let request: NSAsynchronousFetchRequest<NSFetchRequestResult> = cdBorrow(requestPtr)
        let result = try context.execute(request) as? NSAsynchronousFetchResult<NSFetchRequestResult>
        outResult?.pointee = result.map(cdRetain)
        return CDR_OK
    } catch let error as NSError {
        cdWriteError(error, to: outError)
        return Int32(error.code)
    }
}

@_cdecl("cd_persistent_store_async_result_get_managed_object_context")
public func cdPersistentStoreAsyncResultGetManagedObjectContext(_ resultPtr: UnsafeMutableRawPointer?) -> UnsafeMutableRawPointer? {
    guard let resultPtr else {
        return nil
    }
    let result: NSPersistentStoreAsynchronousResult = cdBorrow(resultPtr)
    return cdRetain(result.managedObjectContext)
}

@_cdecl("cd_persistent_store_async_result_get_operation_error_json")
public func cdPersistentStoreAsyncResultGetOperationErrorJSON(_ resultPtr: UnsafeMutableRawPointer?) -> UnsafeMutablePointer<CChar>? {
    guard let resultPtr else {
        return nil
    }
    let result: NSPersistentStoreAsynchronousResult = cdBorrow(resultPtr)
    guard let error = result.operationError as NSError? else {
        return nil
    }
    let payload = CDErrorPayload(domain: error.domain, code: error.code, message: error.localizedDescription)
    return (try? cdEncodeJSON(payload)).flatMap(cdCString)
}

@_cdecl("cd_persistent_store_async_result_get_progress_fraction_completed")
public func cdPersistentStoreAsyncResultGetProgressFractionCompleted(_ resultPtr: UnsafeMutableRawPointer?) -> Double {
    guard let resultPtr else {
        return 0
    }
    let result: NSPersistentStoreAsynchronousResult = cdBorrow(resultPtr)
    return result.progress?.fractionCompleted ?? 0
}

@_cdecl("cd_persistent_store_async_result_has_progress")
public func cdPersistentStoreAsyncResultHasProgress(_ resultPtr: UnsafeMutableRawPointer?) -> Int32 {
    guard let resultPtr else {
        return 0
    }
    let result: NSPersistentStoreAsynchronousResult = cdBorrow(resultPtr)
    return result.progress == nil ? 0 : 1
}

@_cdecl("cd_asynchronous_fetch_result_get_fetch_request")
public func cdAsynchronousFetchResultGetFetchRequest(_ resultPtr: UnsafeMutableRawPointer?) -> UnsafeMutableRawPointer? {
    guard let resultPtr else {
        return nil
    }
    let result: NSAsynchronousFetchResult<NSFetchRequestResult> = cdBorrow(resultPtr)
    return cdRetain(result.fetchRequest)
}

@_cdecl("cd_asynchronous_fetch_result_get_final_result_count")
public func cdAsynchronousFetchResultGetFinalResultCount(_ resultPtr: UnsafeMutableRawPointer?) -> UInt64 {
    guard let resultPtr else {
        return 0
    }
    let result: NSAsynchronousFetchResult<NSFetchRequestResult> = cdBorrow(resultPtr)
    return UInt64(result.finalResult?.count ?? 0)
}

@_cdecl("cd_save_changes_request_new")
public func cdSaveChangesRequestNew(
    _ insertedObjectPtrs: UnsafePointer<UnsafeMutableRawPointer?>?,
    _ insertedCount: Int32,
    _ updatedObjectPtrs: UnsafePointer<UnsafeMutableRawPointer?>?,
    _ updatedCount: Int32,
    _ deletedObjectPtrs: UnsafePointer<UnsafeMutableRawPointer?>?,
    _ deletedCount: Int32,
    _ lockedObjectPtrs: UnsafePointer<UnsafeMutableRawPointer?>?,
    _ lockedCount: Int32,
    _ outRequest: UnsafeMutablePointer<UnsafeMutableRawPointer?>?,
    _ outError: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?
) -> Int32 {
    do {
        let inserted: Set<NSManagedObject> = Set(cdObjects(from: insertedObjectPtrs, count: insertedCount))
        let updated: Set<NSManagedObject> = Set(cdObjects(from: updatedObjectPtrs, count: updatedCount))
        let deleted: Set<NSManagedObject> = Set(cdObjects(from: deletedObjectPtrs, count: deletedCount))
        let locked: Set<NSManagedObject> = Set(cdObjects(from: lockedObjectPtrs, count: lockedCount))
        outRequest?.pointee = cdRetain(NSSaveChangesRequest(
            inserted: inserted.isEmpty ? nil : inserted,
            updated: updated.isEmpty ? nil : updated,
            deleted: deleted.isEmpty ? nil : deleted,
            locked: locked.isEmpty ? nil : locked
        ))
        return CDR_OK
    } catch let error as NSError {
        cdWriteError(error, to: outError)
        return Int32(error.code)
    }
}

private func cdRetainedManagedObjectArray(_ objects: Set<NSManagedObject>?) -> UnsafeMutableRawPointer? {
    guard let objects else {
        return nil
    }
    return cdRetain(Array(objects) as NSArray)
}

@_cdecl("cd_save_changes_request_get_inserted_objects")
public func cdSaveChangesRequestGetInsertedObjects(_ requestPtr: UnsafeMutableRawPointer?) -> UnsafeMutableRawPointer? {
    guard let requestPtr else {
        return nil
    }
    let request: NSSaveChangesRequest = cdBorrow(requestPtr)
    return cdRetainedManagedObjectArray(request.insertedObjects)
}

@_cdecl("cd_save_changes_request_get_updated_objects")
public func cdSaveChangesRequestGetUpdatedObjects(_ requestPtr: UnsafeMutableRawPointer?) -> UnsafeMutableRawPointer? {
    guard let requestPtr else {
        return nil
    }
    let request: NSSaveChangesRequest = cdBorrow(requestPtr)
    return cdRetainedManagedObjectArray(request.updatedObjects)
}

@_cdecl("cd_save_changes_request_get_deleted_objects")
public func cdSaveChangesRequestGetDeletedObjects(_ requestPtr: UnsafeMutableRawPointer?) -> UnsafeMutableRawPointer? {
    guard let requestPtr else {
        return nil
    }
    let request: NSSaveChangesRequest = cdBorrow(requestPtr)
    return cdRetainedManagedObjectArray(request.deletedObjects)
}

@_cdecl("cd_save_changes_request_get_locked_objects")
public func cdSaveChangesRequestGetLockedObjects(_ requestPtr: UnsafeMutableRawPointer?) -> UnsafeMutableRawPointer? {
    guard let requestPtr else {
        return nil
    }
    let request: NSSaveChangesRequest = cdBorrow(requestPtr)
    return cdRetainedManagedObjectArray(request.lockedObjects)
}

@_cdecl("cd_fetch_request_expression_new")
public func cdFetchRequestExpressionNew(
    _ fetchRequestPtr: UnsafeMutableRawPointer?,
    _ contextPtr: UnsafeMutableRawPointer?,
    _ countOnly: Int32,
    _ outExpression: UnsafeMutablePointer<UnsafeMutableRawPointer?>?,
    _ outError: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?
) -> Int32 {
    do {
        guard let fetchRequestPtr, let contextPtr else {
            throw cdBridgeNSError(code: CDR_INVALID_ARGUMENT, message: "Missing fetch request or managed object context")
        }
        let fetchRequest: NSFetchRequest<NSFetchRequestResult> = cdBorrow(fetchRequestPtr)
        let context: NSManagedObjectContext = cdBorrow(contextPtr)
        let expression = NSFetchRequestExpression.expression(
            forFetch: NSExpression(forConstantValue: fetchRequest),
            context: NSExpression(forConstantValue: context),
            countOnly: countOnly != 0
        )
        outExpression?.pointee = cdRetain(expression)
        return CDR_OK
    } catch let error as NSError {
        cdWriteError(error, to: outError)
        return Int32(error.code)
    }
}

@_cdecl("cd_fetch_request_expression_get_count_only_request")
public func cdFetchRequestExpressionGetCountOnlyRequest(_ expressionPtr: UnsafeMutableRawPointer?) -> Int32 {
    guard let expressionPtr else {
        return 0
    }
    let expression: NSFetchRequestExpression = cdBorrow(expressionPtr)
    return expression.isCountOnlyRequest ? 1 : 0
}

File: src/persistent_store_request.rs
use core::ffi::c_void;

use crate::batch_operation::{
    NSBatchDeleteRequest, NSBatchDeleteResult, NSBatchInsertRequest, NSBatchInsertResult,
    NSBatchUpdateRequest, NSBatchUpdateResult,
};
use crate::cloudkit_mirroring::NSPersistentCloudKitContainerEventResult;
use crate::context::{NSManagedObject, NSManagedObjectContext};
use crate::error::CoreDataError;
use crate::ffi;
use crate::history::NSPersistentHistoryResult;
use crate::managed_object::NSManagedObjectID;
use crate::persistent_store_coordinator::NSPersistentStore;
use crate::private::{collect_array, error_from_status, impl_object_wrapper, parse_error_ptr};
use crate::query::NSFetchRequest;

/// Mirrors `NSFetchRequestResult`.
pub trait NSFetchRequestResult {}

impl NSFetchRequestResult for NSManagedObject {}
impl NSFetchRequestResult for NSManagedObjectID {}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
#[non_exhaustive]
/// Mirrors the corresponding Core Data `NSPersistentStoreRequestType` value.
pub enum NSPersistentStoreRequestType {
    /// Mirrors `NSPersistentStoreRequestType::Fetch`.
    Fetch,
    /// Mirrors `NSPersistentStoreRequestType::Save`.
    Save,
    /// Mirrors `NSPersistentStoreRequestType::BatchInsert`.
    BatchInsert,
    /// Mirrors `NSPersistentStoreRequestType::BatchUpdate`.
    BatchUpdate,
    /// Mirrors `NSPersistentStoreRequestType::BatchDelete`.
    BatchDelete,
    /// Mirrors `NSPersistentStoreRequestType::Unknown`.
    Unknown(u64),
}

impl NSPersistentStoreRequestType {
    const fn from_raw(raw: u64) -> Self {
        match raw {
            1 => Self::Fetch,
            2 => Self::Save,
            5 => Self::BatchInsert,
            6 => Self::BatchUpdate,
            7 => Self::BatchDelete,
            other => Self::Unknown(other),
        }
    }
}

impl_object_wrapper!(NSPersistentStoreRequest);
impl_object_wrapper!(NSPersistentStoreResult);
impl_object_wrapper!(NSPersistentStoreAsynchronousResult);
impl_object_wrapper!(NSAsynchronousFetchRequest);
impl_object_wrapper!(NSAsynchronousFetchResult);
impl_object_wrapper!(NSSaveChangesRequest);
impl_object_wrapper!(NSFetchRequestExpression);

fn clone_retained_wrapper<T>(ptr: *mut c_void, context: &str) -> Result<T, CoreDataError>
where
    T: crate::private::FromRetainedPtr,
{
    let retained = unsafe { ffi::cd_retain_object(ptr) };
    unsafe { T::from_retained_ptr(retained, context) }
}

fn collect_object_slice(
    objects: &[&NSManagedObject],
) -> Result<(*const *mut c_void, i32, Vec<*mut c_void>), CoreDataError> {
    let raw_objects = objects
        .iter()
        .map(|object| object.as_ptr())
        .collect::<Vec<_>>();
    let count = i32::try_from(raw_objects.len())
        .map_err(|_| CoreDataError::bridge(-1, "managed object count overflow"))?;
    Ok((raw_objects.as_ptr(), count, raw_objects))
}

impl NSPersistentStoreRequest {
    /// Wraps `NSPersistentStoreRequest.affected_stores(...)`.
    pub fn affected_stores(&self) -> Result<Vec<NSPersistentStore>, CoreDataError> {
        let array_ptr =
            unsafe { ffi::cd_persistent_store_request_get_affected_stores(self.as_ptr()) };
        collect_array(array_ptr, "persistent store request affected stores")
    }

    /// Mirrors `NSPersistentStoreRequest.affected_stores`.
    pub fn set_affected_stores(&self, stores: &[&NSPersistentStore]) -> Result<(), CoreDataError> {
        let raw_stores = stores
            .iter()
            .map(|store| store.as_ptr())
            .collect::<Vec<_>>();
        let count = i32::try_from(raw_stores.len()).map_err(|_| {
            CoreDataError::bridge(-1, "persistent store request store count overflow")
        })?;
        let mut out_error = core::ptr::null_mut();
        let status = unsafe {
            ffi::cd_persistent_store_request_set_affected_stores(
                self.as_ptr(),
                raw_stores.as_ptr(),
                count,
                &mut out_error,
            )
        };
        if status != ffi::status::OK {
            return Err(unsafe { error_from_status(status, out_error) });
        }
        Ok(())
    }

    /// Wraps `NSPersistentStoreRequest.request_type(...)`.
    pub fn request_type(&self) -> NSPersistentStoreRequestType {
        NSPersistentStoreRequestType::from_raw(unsafe {
            ffi::cd_persistent_store_request_get_request_type(self.as_ptr())
        })
    }
}

impl NSPersistentStoreResult {
    /// Wraps `NSPersistentStoreResult.as_asynchronous_result(...)`.
    pub fn as_asynchronous_result(
        &self,
    ) -> Result<NSPersistentStoreAsynchronousResult, CoreDataError> {
        clone_retained_wrapper(self.as_ptr(), "persistent store asynchronous result")
    }
}

impl NSPersistentStoreAsynchronousResult {
    /// Wraps `NSPersistentStoreAsynchronousResult.managed_object_context(...)`.
    pub fn managed_object_context(&self) -> Result<NSManagedObjectContext, CoreDataError> {
        let ptr = unsafe {
            ffi::cd_persistent_store_async_result_get_managed_object_context(self.as_ptr())
        };
        unsafe {
            NSManagedObjectContext::from_retained_ptr(
                ptr,
                "persistent store asynchronous result context",
            )
        }
    }

    /// Wraps `NSPersistentStoreAsynchronousResult.operation_error(...)`.
    pub fn operation_error(&self) -> Option<CoreDataError> {
        let ptr = unsafe {
            ffi::cd_persistent_store_async_result_get_operation_error_json(self.as_ptr())
        };
        if ptr.is_null() {
            return None;
        }
        Some(unsafe { parse_error_ptr(ptr) })
    }

    /// Wraps `NSPersistentStoreAsynchronousResult.progress_fraction_completed(...)`.
    pub fn progress_fraction_completed(&self) -> Option<f64> {
        if unsafe { ffi::cd_persistent_store_async_result_has_progress(self.as_ptr()) } == 0 {
            return None;
        }
        Some(unsafe {
            ffi::cd_persistent_store_async_result_get_progress_fraction_completed(self.as_ptr())
        })
    }
}

impl NSAsynchronousFetchRequest {
    /// Wraps `NSAsynchronousFetchRequest.init(...)`.
    pub fn new(fetch_request: &NSFetchRequest) -> Result<Self, CoreDataError> {
        let mut out_request = core::ptr::null_mut();
        let mut out_error = core::ptr::null_mut();
        let status = unsafe {
            ffi::cd_asynchronous_fetch_request_new(
                fetch_request.as_ptr(),
                &mut out_request,
                &mut out_error,
            )
        };
        if status != ffi::status::OK {
            return Err(unsafe { error_from_status(status, out_error) });
        }
        unsafe { Self::from_retained_ptr(out_request, "asynchronous fetch request") }
    }

    /// Wraps `NSAsynchronousFetchRequest.fetch_request(...)`.
    pub fn fetch_request(&self) -> Result<NSFetchRequest, CoreDataError> {
        let ptr = unsafe { ffi::cd_asynchronous_fetch_request_get_fetch_request(self.as_ptr()) };
        unsafe {
            NSFetchRequest::from_retained_ptr(ptr, "asynchronous fetch request fetch request")
        }
    }

    /// Wraps `NSAsynchronousFetchRequest.estimated_result_count(...)`.
    pub fn estimated_result_count(&self) -> i64 {
        unsafe { ffi::cd_asynchronous_fetch_request_get_estimated_result_count(self.as_ptr()) }
    }

    /// Mirrors `NSAsynchronousFetchRequest.estimated_result_count`.
    pub fn set_estimated_result_count(&self, estimated_result_count: i64) {
        unsafe {
            ffi::cd_asynchronous_fetch_request_set_estimated_result_count(
                self.as_ptr(),
                estimated_result_count,
            );
        }
    }

    /// Wraps `NSAsynchronousFetchRequest.execute(...)`.
    pub fn execute(
        &self,
        context: &NSManagedObjectContext,
    ) -> Result<NSAsynchronousFetchResult, CoreDataError> {
        let mut out_result = core::ptr::null_mut();
        let mut out_error = core::ptr::null_mut();
        let status = unsafe {
            ffi::cd_managed_object_context_execute_asynchronous_fetch_request(
                context.as_ptr(),
                self.as_ptr(),
                &mut out_result,
                &mut out_error,
            )
        };
        if status != ffi::status::OK {
            return Err(unsafe { error_from_status(status, out_error) });
        }
        unsafe {
            NSAsynchronousFetchResult::from_retained_ptr(out_result, "asynchronous fetch result")
        }
    }

    /// Wraps `NSAsynchronousFetchRequest.as_persistent_store_request(...)`.
    pub fn as_persistent_store_request(&self) -> Result<NSPersistentStoreRequest, CoreDataError> {
        clone_retained_wrapper(self.as_ptr(), "persistent store request")
    }
}

impl NSAsynchronousFetchResult {
    /// Wraps `NSAsynchronousFetchResult.asynchronous_result(...)`.
    pub fn asynchronous_result(
        &self,
    ) -> Result<NSPersistentStoreAsynchronousResult, CoreDataError> {
        clone_retained_wrapper(self.as_ptr(), "persistent store asynchronous result")
    }

    /// Wraps `NSAsynchronousFetchResult.fetch_request(...)`.
    pub fn fetch_request(&self) -> Result<NSAsynchronousFetchRequest, CoreDataError> {
        let ptr = unsafe { ffi::cd_asynchronous_fetch_result_get_fetch_request(self.as_ptr()) };
        unsafe {
            NSAsynchronousFetchRequest::from_retained_ptr(ptr, "asynchronous fetch result request")
        }
    }

    /// Wraps `NSAsynchronousFetchResult.final_result_count(...)`.
    pub fn final_result_count(&self) -> usize {
        unsafe { ffi::cd_asynchronous_fetch_result_get_final_result_count(self.as_ptr()) as usize }
    }
}

impl NSSaveChangesRequest {
    /// Wraps `NSSaveChangesRequest.init(...)`.
    pub fn new(
        inserted_objects: &[&NSManagedObject],
        updated_objects: &[&NSManagedObject],
        deleted_objects: &[&NSManagedObject],
        locked_objects: &[&NSManagedObject],
    ) -> Result<Self, CoreDataError> {
        let (inserted_ptr, inserted_count, inserted_raw) = collect_object_slice(inserted_objects)?;
        let (updated_ptr, updated_count, updated_raw) = collect_object_slice(updated_objects)?;
        let (deleted_ptr, deleted_count, deleted_raw) = collect_object_slice(deleted_objects)?;
        let (locked_ptr, locked_count, locked_raw) = collect_object_slice(locked_objects)?;
        let mut out_request = core::ptr::null_mut();
        let mut out_error = core::ptr::null_mut();
        let status = unsafe {
            ffi::cd_save_changes_request_new(
                inserted_ptr,
                inserted_count,
                updated_ptr,
                updated_count,
                deleted_ptr,
                deleted_count,
                locked_ptr,
                locked_count,
                &mut out_request,
                &mut out_error,
            )
        };
        let _ = (inserted_raw, updated_raw, deleted_raw, locked_raw);
        if status != ffi::status::OK {
            return Err(unsafe { error_from_status(status, out_error) });
        }
        unsafe { Self::from_retained_ptr(out_request, "save changes request") }
    }

    /// Wraps `NSSaveChangesRequest.inserted_objects(...)`.
    pub fn inserted_objects(&self) -> Result<Vec<NSManagedObject>, CoreDataError> {
        let array_ptr = unsafe { ffi::cd_save_changes_request_get_inserted_objects(self.as_ptr()) };
        collect_array(array_ptr, "save changes request inserted objects")
    }

    /// Wraps `NSSaveChangesRequest.updated_objects(...)`.
    pub fn updated_objects(&self) -> Result<Vec<NSManagedObject>, CoreDataError> {
        let array_ptr = unsafe { ffi::cd_save_changes_request_get_updated_objects(self.as_ptr()) };
        collect_array(array_ptr, "save changes request updated objects")
    }

    /// Wraps `NSSaveChangesRequest.deleted_objects(...)`.
    pub fn deleted_objects(&self) -> Result<Vec<NSManagedObject>, CoreDataError> {
        let array_ptr = unsafe { ffi::cd_save_changes_request_get_deleted_objects(self.as_ptr()) };
        collect_array(array_ptr, "save changes request deleted objects")
    }

    /// Wraps `NSSaveChangesRequest.locked_objects(...)`.
    pub fn locked_objects(&self) -> Result<Vec<NSManagedObject>, CoreDataError> {
        let array_ptr = unsafe { ffi::cd_save_changes_request_get_locked_objects(self.as_ptr()) };
        collect_array(array_ptr, "save changes request locked objects")
    }

    /// Wraps `NSSaveChangesRequest.as_persistent_store_request(...)`.
    pub fn as_persistent_store_request(&self) -> Result<NSPersistentStoreRequest, CoreDataError> {
        clone_retained_wrapper(self.as_ptr(), "persistent store request")
    }
}

impl NSFetchRequestExpression {
    /// Wraps `NSFetchRequestExpression.init(...)`.
    pub fn new(
        fetch_request: &NSFetchRequest,
        context: &NSManagedObjectContext,
        count_only: bool,
    ) -> Result<Self, CoreDataError> {
        let mut out_expression = core::ptr::null_mut();
        let mut out_error = core::ptr::null_mut();
        let status = unsafe {
            ffi::cd_fetch_request_expression_new(
                fetch_request.as_ptr(),
                context.as_ptr(),
                i32::from(count_only),
                &mut out_expression,
                &mut out_error,
            )
        };
        if status != ffi::status::OK {
            return Err(unsafe { error_from_status(status, out_error) });
        }
        unsafe { Self::from_retained_ptr(out_expression, "fetch request expression") }
    }

    /// Wraps `NSFetchRequestExpression.is_count_only_request(...)`.
    pub fn is_count_only_request(&self) -> bool {
        unsafe { ffi::cd_fetch_request_expression_get_count_only_request(self.as_ptr()) != 0 }
    }
}

impl NSFetchRequest {
    /// Wraps `NSFetchRequest.as_persistent_store_request(...)`.
    pub fn as_persistent_store_request(&self) -> Result<NSPersistentStoreRequest, CoreDataError> {
        clone_retained_wrapper(self.as_ptr(), "persistent store request")
    }
}

impl NSBatchDeleteRequest {
    /// Wraps `NSBatchDeleteRequest.as_persistent_store_request(...)`.
    pub fn as_persistent_store_request(&self) -> Result<NSPersistentStoreRequest, CoreDataError> {
        clone_retained_wrapper(self.as_ptr(), "persistent store request")
    }
}

impl NSBatchInsertRequest {
    /// Wraps `NSBatchInsertRequest.as_persistent_store_request(...)`.
    pub fn as_persistent_store_request(&self) -> Result<NSPersistentStoreRequest, CoreDataError> {
        clone_retained_wrapper(self.as_ptr(), "persistent store request")
    }
}

impl NSBatchUpdateRequest {
    /// Wraps `NSBatchUpdateRequest.as_persistent_store_request(...)`.
    pub fn as_persistent_store_request(&self) -> Result<NSPersistentStoreRequest, CoreDataError> {
        clone_retained_wrapper(self.as_ptr(), "persistent store request")
    }
}

impl NSBatchDeleteResult {
    /// Wraps `NSBatchDeleteResult.as_persistent_store_result(...)`.
    pub fn as_persistent_store_result(&self) -> Result<NSPersistentStoreResult, CoreDataError> {
        clone_retained_wrapper(self.as_ptr(), "persistent store result")
    }
}

impl NSBatchInsertResult {
    /// Wraps `NSBatchInsertResult.as_persistent_store_result(...)`.
    pub fn as_persistent_store_result(&self) -> Result<NSPersistentStoreResult, CoreDataError> {
        clone_retained_wrapper(self.as_ptr(), "persistent store result")
    }
}

impl NSBatchUpdateResult {
    /// Wraps `NSBatchUpdateResult.as_persistent_store_result(...)`.
    pub fn as_persistent_store_result(&self) -> Result<NSPersistentStoreResult, CoreDataError> {
        clone_retained_wrapper(self.as_ptr(), "persistent store result")
    }
}

impl NSPersistentHistoryResult {
    /// Wraps `NSPersistentHistoryResult.as_persistent_store_result(...)`.
    pub fn as_persistent_store_result(&self) -> Result<NSPersistentStoreResult, CoreDataError> {
        clone_retained_wrapper(self.as_ptr(), "persistent store result")
    }
}

impl NSPersistentCloudKitContainerEventResult {
    /// Wraps `NSPersistentCloudKitContainerEventResult.as_persistent_store_result(...)`.
    pub fn as_persistent_store_result(&self) -> Result<NSPersistentStoreResult, CoreDataError> {
        clone_retained_wrapper(self.as_ptr(), "persistent store result")
    }
}
